### ISSUE TITLE
support slice indexing by field projection

### ISSUE THREAD

**Issue by timotheecour** [libmir/mir-algorithm#54]
i had implemented slice indexing by struct field projection in my own tensor library. Would be great to support that in mir.

Here's the idea:

```
struct A{  int x;   string name; }
auto s=3.iota.map!(a=>A(a,"foo")).array.slice;

assert(s.fields.name==["foo", "foo"]);

// s.fields.x has a stride and backed by int*; it doesnt' allocate.
s.fields.x[]+=10;

assert(s[0]==10);
```

Note: this is more powerful than `s.map!(a=>a.x)` since the latter wouldn't be a slice backed by a `int*` (and in particular wouldn't contain lvalues).

## implementation
N-dimensional field projection is just an indexing operation when treating A as `void[0..A.sizeof]`:
`s.fields.x` is:
`s.field.Cast!(void[]).sliced(s.length, A.sizeof)[0..$, 0..A.init.x.sizeof].reinterpretCast!int;`

with:
```
auto Cast(T, S)(S a){ return cast(T)a;}
```

Side question: is there an ndslice reinterpretCast ? it would be the slice analog of `cast(T[]) a` which just casts an array of S as an array of T without changing the bits.

## NOTE
using directly opDispatch on Slice is too dangerous as it could lead to name clashes between Slice's existing fields/methods and the elements, hence we use an intermediate `fields` which then forwards to opDispatch. Also, not sure if `fields` is the best name, as it's completely different meaning than `field`.

another possible syntax:`s.fields!"x"` but maybe less nice-looking.

**Comment by 9il** [comment]
This kind of functionality can be included. Will you open PR?

> is there an ndslice reinterpretCast ?

ndslice has not it

**Comment by 9il** [comment]
First implementation step: slice should be converted to universal kind.

**Comment by 9il** [comment]
name can agreed later when PR is ready

**Comment by 9il** [comment]
functions should be located in `mir.ndslice.topology`

**Comment by 9il** [comment]
Fixed in 2b7970ac849a5cdd272b32e2ddbb4655d93c558c
